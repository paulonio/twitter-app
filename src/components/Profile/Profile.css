.wrapper {
  width: 100%;
  max-width: 1580px;
  margin: 0 auto;
}