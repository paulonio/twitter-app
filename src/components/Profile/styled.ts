import { styled } from 'styled-components';

export const PageWrapper = styled.div`
  width: 100%;
  max-width: 1580px;
  margin: 0 auto;
  display: flex;
  justify-content: space-between;

  @media screen and (max-width: 1480px) {
    justify-content: center;
    gap: 24px;
  }
`;

export const ProfileWrapper = styled.section`
  width: 57.5%;
  max-width: 910px;
  min-height: 100vh;
  border-left: solid 1px ${({ theme }) => theme.colors.border};
  border-right: solid 1px ${({ theme }) => theme.colors.border};

  @media screen and (max-width: 1480px) {
    width: 70%;
  }

  @media screen and (max-width: 768px) {
    width: 100%;
    overflow: hidden;
  }
`;

export const HeaderTitle = styled.div`
  font-size: 20px;
  line-height: 27px;
  font-weight: 700;
  color: ${({ theme }) => theme.colors.text};
`;

export const HeaderSubtitle = styled.div`
  font-size: 16px;
  line-height: 21px;
  font-weight: 400;
  color: ${({ theme }) => theme.colors.text};
  opacity: 0.6;
`;

export const ProfileImage = styled.img.attrs({
  src: '/assets/images/Profile-image.jpg',
  alt: 'Profile image',
})`
  width: 100%;
  height: auto;
  margin-top: 85px;
`;

export const UserWrapper = styled.div`
  width: 100%;
  max-width: 910px;
  padding: 0 25px 48px 25px;
  transform: translateY(-15%);
  display: flex;
  flex-direction: column;
  align-items: start;
  justify-content: center;

  @media screen and (max-width: 470px) {
    padding: 0 8px 24px 8px;
  }
`;

export const Username = styled.h2`
  font-size: 24px;
  line-height: 32px;
  font-weight: 700;
  color: ${({ theme }) => theme.colors.text};

  @media screen and (max-width: 470px) {
    font-size: 16px;
    line-height: 21px;
  }
`;

export const Nickname = styled.div`
  font-size: 16px;
  line-height: 21px;
  font-weight: 400;
  color: ${({ theme }) => theme.colors.text};
  opacity: 0.6;
`;

export const UserBio = styled.p`
  margin-top: 15px;
  font-size: 18px;
  line-height: 24px;
  font-weight: 400;
  color: ${({ theme }) => theme.colors.text};
`;

export const AvatarRow = styled.div`
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

export const AvatarWrapper = styled.div`
  height: 150px;
  width: 150px;

  @media screen and (max-width: 768px) {
    width: 75px;
    height: 75px;
  }
`;

export const ButtonWrapper = styled.div`
  transform: translateY(50%);
`;

export const FollowersWrapper = styled.div`
  margin-top: 54px;
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

export const FollowersItem = styled.div`
  margin-left: 30px;
  font-size: 18px;
  line-height: 24px;
  font-weight: 400;
  color: ${({ theme }) => theme.colors.text};
  opacity: 0.6;
  &:first-child {
    margin-left: 0;
  }

  @media screen and (max-width: 470px) {
    font-size: 16px;
    line-height: 21px;
  }
`;

export const Amount = styled.span`
  color: ${({ theme }) => theme.colors.text};
  opacity: 1;
  font-weight: 700;
`;
