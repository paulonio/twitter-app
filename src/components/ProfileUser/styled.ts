import { styled } from 'styled-components';

export const UserWrapper = styled.div`
  width: 100%;
  max-width: 910px;
  padding: 0 25px 48px 25px;
  transform: translateY(-15%);
  display: flex;
  flex-direction: column;
  align-items: start;
  justify-content: center;
`;

export const Username = styled.h2`
  font-size: 24px;
  line-height: 32px;
  font-weight: 700;
`;

export const Nickname = styled.div`
  font-size: 16px;
  line-height: 21px;
  font-weight: 400;
  color: rgba(0, 0, 0, 0.6);
`;

export const UserBio = styled.p`
  margin-top: 15px;
  font-size: 18px;
  line-height: 24px;
  font-weight: 400;
`;
