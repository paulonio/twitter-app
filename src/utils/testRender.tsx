import React, { FC } from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import type { PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import Theme from '@src/components/Theme/Theme';
import { persistor } from '@src/store';
import { BrowserRouter } from 'react-router-dom';
import { PersistGate } from 'redux-persist/integration/react';
import { setupStore, AppStore, RootState } from './testUtils';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

interface WrapperProps {
  children: JSX.Element | JSX.Element[] | string | string[];
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  const Wrapper: FC<WrapperProps> = ({ children }): JSX.Element => {
    return (
      <Provider store={store}>
        <PersistGate loading={null} persistor={persistor}>
          <BrowserRouter>
            <Theme>{children}</Theme>
          </BrowserRouter>
        </PersistGate>
      </Provider>
    );
  };
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};
