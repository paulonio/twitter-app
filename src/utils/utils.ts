import {
  doc,
  DocumentSnapshot,
  DocumentData,
  getDoc,
  collection,
  updateDoc,
  arrayUnion,
} from 'firebase/firestore';
import { db } from '../../firebase';
import { TweetType } from '../store/slices/tweetSlice';
import { User } from '../store/slices/authSlice';

export const getDocument = async (column: string, document: string) => {
  const docRef = doc(db, column, document);
  const docSnap: DocumentSnapshot<DocumentData> = await getDoc(docRef);
  return docSnap;
};

export const updateDocument = async (column: string, document: string, payload: TweetType) => {
  const twitterRef = collection(db, column);
  await updateDoc(doc(twitterRef, document), { tweets: arrayUnion(payload) });
  const docSnap: DocumentSnapshot<DocumentData> = await getDocument(column, document);
  return docSnap;
};

export const updateUsers = async (column: string, document: string, payload: User | User[]) => {
  const twitterRef = collection(db, column);
  if (payload instanceof Array) {
    await updateDoc(doc(twitterRef, document), { users: payload });
  } else {
    await updateDoc(doc(twitterRef, document), { users: arrayUnion(payload) });
  }
  const docSnap: DocumentSnapshot<DocumentData> = await getDocument(column, document);
  return docSnap;
};

export const isUserArray = (value: unknown): value is User[] => {
  if (!Array.isArray(value)) {
    return false;
  }

  return value.some((item) => !!item.uid);
};

export const isTweetArray = (value: unknown): value is TweetType[] => {
  if (!Array.isArray(value)) {
    return false;
  }

  return value.some((item) => !!item.tweet);
};
