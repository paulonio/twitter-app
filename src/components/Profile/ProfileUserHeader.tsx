import React, { useState } from 'react';

import Avatar from '@components/Avatar/Avatar';
import EditUser from '@components/EditUser/EditUser';
import Button from '@components/Button/Button';

import { AvatarRow, AvatarWrapper, ButtonWrapper } from './styled';

const ProfileUserHeader = () => {
  const [isActive, setActive] = useState<boolean>(false);

  const handleOpenModal = () => {
    setActive(true);
  };

  return (
    <AvatarRow>
      <EditUser isActive={isActive} setActive={setActive} />
      <AvatarWrapper>
        <Avatar />
      </AvatarWrapper>
      <ButtonWrapper>
        <Button $buttonType="bordered" onClick={handleOpenModal}>
          Edit profile
        </Button>
      </ButtonWrapper>
    </AvatarRow>
  );
};

export default ProfileUserHeader;
