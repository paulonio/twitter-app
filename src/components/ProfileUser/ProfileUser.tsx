import React from 'react';
import { useSelector } from 'react-redux';
import { User } from 'firebase/auth';
import { Nickname, UserBio, UserWrapper, Username } from './styled';
import Followers from '../Followers/Followers';
import ProfileUserHeader from '../ProfileUserHeader/ProfileUserHeader';
import { StoreType } from '../../store';

const ProfileUser = () => {
  const user = useSelector<StoreType, User | null>((state) => state.auth.user);

  return (
    <UserWrapper>
      <ProfileUserHeader />
      <Username>{user?.displayName || 'Bobur'}</Username>
      <Nickname>@{user?.email?.split('@')[0] || '@bobur'}</Nickname>
      <UserBio>UX&UI designer at @abutechuz</UserBio>
      <Followers />
    </UserWrapper>
  );
};

export default ProfileUser;
