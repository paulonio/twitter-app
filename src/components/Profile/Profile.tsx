import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Navigate } from 'react-router-dom';

import Header from '@components/Header/Header';
import ProfileUser from '@src/components/Profile/ProfileUser';
import TweetBlock from '@components/TweetBlock/TweetBlock';
import Tweets from '@components/Tweets/Tweets';
import ErrorBoundary from '@components/ErrorBoundary/ErrorBoundary';

import type { StoreType } from '@store/index.ts';
import type { UserType } from '@store/slices/authSlice';
import type { TweetType } from '@store/slices/tweetSlice';
import { syncTweetsRequest } from '@store/actions/actions';

import { HeaderSubtitle, HeaderTitle, ProfileImage, ProfileWrapper } from './styled';

const Profile = () => {
  const user = useSelector<StoreType, UserType | null>((state) => state.auth.user);
  const tweets = useSelector<StoreType, TweetType[]>((state) => state.tweet.profileTweets);
  const dispatch = useDispatch();
  const profileTweetsAmount = tweets.length;

  useEffect(() => {
    if (user) {
      dispatch(syncTweetsRequest(user));
    }
  }, []);

  if (!user) {
    return <Navigate to="/" replace />;
  }

  return (
    <ProfileWrapper>
      <ErrorBoundary>
        <Header>
          <HeaderTitle>{user?.displayName}</HeaderTitle>
          <HeaderSubtitle>{profileTweetsAmount} Tweets</HeaderSubtitle>
        </Header>
        <ProfileImage />
        <ProfileUser />
        <TweetBlock />
        <Tweets type="profile" />
      </ErrorBoundary>
    </ProfileWrapper>
  );
};

export default Profile;
