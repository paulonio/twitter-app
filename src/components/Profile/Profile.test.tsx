import React from 'react';
import { screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import Profile from '@components/Profile/Profile';
import ProfileUser from '@components/Profile/ProfileUser';

import { renderWithProviders } from '@src/utils/testRender';
import type { RootState } from '@src/utils/testUtils';

export const testData: RootState = {
  auth: {
    user: { email: 'test@test.com', displayName: 'test', uid: 'test' },
    allUsers: [],
    error: null,
  },
  tweet: { tweets: [], profileTweets: [], error: null },
  theme: { theme: 'dark' },
};

describe('Profile', () => {
  it('should render profile component', () => {
    const { container } = renderWithProviders(<Profile />);

    expect(container).toMatchSnapshot();
  });

  it('should render profile user correctly', () => {
    renderWithProviders(<ProfileUser />, { preloadedState: testData });

    const username = screen.getByTestId('username');
    const nickname = screen.getByTestId('nickname');

    expect(username.textContent).toEqual('test');
    expect(nickname.textContent).toEqual('@test');
  });
});
