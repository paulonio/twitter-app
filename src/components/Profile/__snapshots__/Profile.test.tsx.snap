// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile should render profile component 1`] = `<div />`;
